### ISSUE TITLE
Memory leak in Client

### ISSUE THREAD

**Issue by everhardt** [patrickjuchli/basic-ftp#249]
**Describe the bug**
Since using basic-ftp I had a memory leak in my project. I managed to reproduce it by the following sample code, that does nothing else than connecting and disconnecting. If I plot `heapUsed` and ignore the first ~80 points because of initialisation, you see through the jigsaw of garbage collection that it slowly but steadily increases.

<img width="680" alt="image" src="https://github.com/patrickjuchli/basic-ftp/assets/1840153/27d7a7a9-5080-418a-a6dc-933266e04a27">

**Example code**
```js
const ftp = require('basic-ftp');

async function go() {
  while (true) {
    const before = process.memoryUsage();
    const memUsage = ['heapUsed', 'external', 'arrayBuffers']
      .map((prop) => {
        const value = before[prop] / 1024 / 1024;
        return `${prop}: ${value.toFixed(1)} MiB`;
      })
      .join(', ');

    console.log(memUsage);

    const client = new ftp.Client(5000);

    await client.access({
      host: '127.0.0.1',
      password: 'test',
      user: 'test',
      port: 21,
    });

    client.close();

    await new Promise((resolve) => {
      setTimeout(resolve, 100);
    });
  }
}

go()
  .then(() => {
    process.exit(0);
  })
  .catch((e) => {
    console.log(e);
    process.exit(1);
  });
```

**Console output**
```
Connected to 127.0.0.1:21 (No encryption)
< 220 Ready

> OPTS UTF8 ON
< 530 Command requires authentication: OPTS

Login security: No encryption
> USER test
< 331 Username okay, awaiting password

> PASS ###
< 230 User logged in, proceed

> FEAT
< 211-Extensions supported

<  AUTH TLS
 MDTM

<  PBSZ
 PROT
 SIZE
 TYPE A,I,L
 UTF8
211 End

> TYPE I
< 200 Switch to "binary" transfer mode.

> STRU F
< 200 The requested action has been successfully completed

> OPTS UTF8 ON
< 200 UTF8 encoding on

> QUIT
```

**Which version of Node.js are you using?**
v18.19.0

**Which version of basic-ftp are you using?**
5.0.4

**Additional context**
Tried it with multiple FTP servers, on multiple OS'es (macos and linux) and both with and without TLS.

**Comment by patrickjuchli** [comment]
Thanks for the ticket. I had a quick look using the allocation sampler of the memory profiler but the only thing that I see right away is some Node internals. See screenshot, blue bars represent data not collected by GC yet. I've run this without your code-based memory usage tracking and logging.

<img width="1006" alt="memory" src="https://github.com/patrickjuchli/basic-ftp/assets/397839/f69f361e-5fe6-4baa-a5a3-93eae9e79b82">

It's not immediately obvious to me what this is and whether its something the library is responsible for. Could you have a look on your end? I can see entries related to timers but nothing that the library itself does. It doesn't use any `setTimeout` for example but uses internal timeout tracking of `Socket` instances.

I've used `node --inspect myTest.js`, then opened Chrome with `chrome://inspect`, selected the remote target `myTest.js`. Then in the `Memory` tab I've chosen "Allocation Sampling" and let it run for a bit.

**Comment by everhardt** [comment]
I made 2 heapsnapshots, one at iteration 102 of the while loop and the other at iteration 224. Given the graph above the garbage collector must have run multiple times in between. I compared the snapshots and see that 122 Socket objects are retained:

<img width="1440" alt="image" src="https://github.com/patrickjuchli/basic-ftp/assets/1840153/34545ce8-f16d-4546-a8f4-932c716c3f19">

I don't understand the Retainers overview yet, but I do see `timeout` being listed. I can't pinpoint the problem yet, but tried already with `setTimeout(0)` without any difference.

**Comment by patrickjuchli** [comment]
Are you sure that isn't just the total number of sockets ever allocated?

**Comment by patrickjuchli** [comment]
Your screenshot shows a "summary", not a "comparison" between snapshots as the screenshot here below. See the summary/comparison dropdown at the top left. I haven't run as many iterations as you do so, and I'm absolutely open to the possibility of a memory leak. But I haven't seen anything specific yet that originates from the library.

<img width="1006" alt="Screenshot 2024-02-17 at 20 11 41" src="https://github.com/patrickjuchli/basic-ftp/assets/397839/16d390d7-04c9-4509-9a43-4e6576d0d6e4">

**Comment by everhardt** [comment]
I'm off for a week, but will investigate further after that. 

Ps. My screenshot indeed shows the summary, I selected only objects allocated between both snapshots (see a bit to the right). I would say that it cannot know about objects that are removed in between, as these are snapshots, not captures of the whole time window.

**Comment by patrickjuchli** [comment]
Yes, to see garbage collection in action you need to use the allocation sampler of my first reply or use the comparison tool to compare single snapshots to see what was removed (see columns referring to deleted, freed and deltas).

In any case, thanks for your help!

**Comment by everhardt** [comment]
I've now also used `--inspect` and chose "Allocation instrumentation on timeline" as it states "Use this profile type to isolate memory leaks" and I checked "Record stack traces of allocations". Then I see similar stuff as in my screenshot above, ao. a lot of Socket objects (and FTPContext objects):


<img width="1439" alt="image" src="https://github.com/patrickjuchli/basic-ftp/assets/1840153/49eb8d4d-9e09-430b-8cc1-d715fba056e2">

**Comment by martijnimhoff** [comment]
Also chiming in to help out. 

I did a little bit of digging and this is what I found about net.Socket: https://stackoverflow.com/a/33871679/5688047
> Nodejs caches the Socket objects to be used in future.

Perhaps that is what is happening? I need to dive in a bit more to understand what is happening.

**Comment by martijnimhoff** [comment]
In FtpContext.ts in _closeSocket I found that `socket.destroy()` on line 395 doesn't happen. I tried to add `socket.destroy()` just below `socket.end()` on line 397 and this results in a memory profile without Socket objects. 

Why was that timeout needed in the first place?

**Comment by patrickjuchli** [comment]
I'm okay with merging the suggested PR. The whole thing with the timeout was probably me way overthinking something. @everhardt any last words? If you try the suggested PR do you still see the issues on your end?

**Comment by everhardt** [comment]
I can confirm that the memory leak is gone with the suggested PR. Thanks @martijnimhoff and @patrickjuchli!

**Comment by patrickjuchli** [comment]
Fix is published in 5.0.5
